src/bitcoin-cli.cpp

```diff
@@ -291,6 +295,211 @@ class GetinfoRequestHandler: public BaseRequestHandler
+    };
+    struct Peer {
+        int id;
+        int mapped_as;
+        int version;
+        int64_t conn_time;
+        int64_t last_blck;
+        int64_t last_recv;
+        int64_t last_send;
+        int64_t last_trxn;
+        double min_ping;
+        double ping;
+        std::string addr;
+        std::string sub_version;
+        NetType net_type;
+        bool is_block_relay;
+        bool is_outbound;
+        bool operator<(const Peer& rhs) const { return std::tie(is_outbound, min_ping) < std::tie(rhs.is_outbound, rhs.min_ping); }
+    };
+    std::string NetTypeEnumToString(NetType t)
+    {
+        switch (t) {
+        case NetType::ipv4: return "ipv4";
+        case NetType::ipv6: return "ipv6";
+        case NetType::onion: return "onion";
+        } // no default case, so the compiler can warn about missing cases
+        assert(false);
+    }
+    std::string ChainToString() const
+    {
+        if (gArgs.GetChainName() == CBaseChainParams::TESTNET) return " testnet";
+        if (gArgs.GetChainName() == CBaseChainParams::REGTEST) return " regtest";
+        return "";
+    }
+public:
+    const int ID_PEERINFO = 0;
+    const int ID_NETWORKINFO = 1;
+
+    UniValue PrepareRequest(const std::string& method, const std::vector<std::string>& args) override
+    {
+        if (!args.empty()) {
+            uint8_t n{0};
+            if (ParseUInt8(args.at(0), &n)) {
+                m_details_level = n;
+            }
+        }
+        UniValue result(UniValue::VARR);
+        result.push_back(JSONRPCRequestObj("getpeerinfo", NullUniValue, ID_PEERINFO));
+        result.push_back(JSONRPCRequestObj("getnetworkinfo", NullUniValue, ID_NETWORKINFO));
+        return result;
+    }
+
+    UniValue ProcessReply(const UniValue& batch_in) override
+    {
+        const std::vector<UniValue> batch{JSONRPCProcessBatchReply(batch_in)};
+        if (!batch[ID_PEERINFO]["error"].isNull()) return batch[ID_PEERINFO];
+        if (!batch[ID_NETWORKINFO]["error"].isNull()) return batch[ID_NETWORKINFO];
+
+        const UniValue& networkinfo{batch[ID_NETWORKINFO]["result"]};
+        if (networkinfo["version"].get_int() < 209900) {
+            throw std::runtime_error("-netinfo requires bitcoind server to be running v0.21.0 and up");
+        }
+
+        // Count peer connection totals, and if DetailsRequested(), store peer data in a vector of structs.
+        const int64_t time_now{GetSystemTimeInSeconds()};
+        int ipv4_i{0}, ipv6_i{0}, onion_i{0}, block_relay_i{0}, total_i{0}; // inbound conn counters
+        int ipv4_o{0}, ipv6_o{0}, onion_o{0}, block_relay_o{0}, total_o{0}; // outbound conn counters
+        size_t max_peer_id_length{2}, max_addr_length{0};
+        bool is_asmap_on{false};
+        std::vector<Peer> peers;
+        const UniValue& getpeerinfo{batch[ID_PEERINFO]["result"]};
+
+        for (const UniValue& peer : getpeerinfo.getValues()) {
+            const std::string addr{peer["addr"].get_str()};
```

Comment:
`addr` can also contain hostnames (`bitcoin-cli addnode "foo.bar" onetry`) or IPv6 addresses without square brackets (`bitcoin-cli addnode "2001:db8::1337" onetry`). In theses cases the `IsAddrIPv6` will not recognize them properly and they will be always marked as IPv4 🤔.

Reply:
Maybe, if there are no square brackets, ask `inet_pton` if it´s a valid IPv6 address? But what if there´s a hostname?

Reply:
Dug a little bit deeper: Adding a new field to `getpeerinfo` seems to be a better solution IMO. The actual remote address (that is IPv6 with square brackets or resolved hostname) is already stored [here](https://github.com/bitcoin/bitcoin/blob/master/src/net.h#L687), but only `addrName` is [passed back](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/net.cpp#L161) as `addr` instead.
I would suggest to return both `addrName` & `addr` so that the former contains the hostname (or ordinary IP address) and the latter the actual resolved remote address. This would at least solve the hostname/IPv6 problems, not sure how that effects onion addresses. Let me know what you think and if I should prepare a PR :)

Reply:
@n-thumann Thanks for having a look. I don't plan to make further changes for edge cases unless they can be addressed within the scope of this PR. The net address code is currently being deeply changed for BIP155 `addrv2`, this PR has seen a lot of review and updates, and if we begin gating it on changes in code outside this PR, we'll miss the next feature freeze. This is *very* useable in its current state (if you start using it, good luck doing without it afterward :smiley:) and if merged, I'll be happy to make further improvements based on the state of the codebase at that time.

Reply:
@n-thumann In the interim, don't hesitate to use this feature to keep tabs on your node's peer connections and ACK if you find it satisfactory.